core/shared/src/main/scala/scalaz/zio/Semaphore.scala

```diff
@@ -4,112 +4,69 @@
-              else Right(available - requested)
-            // n in queue but no more to release
-            case nonEmptyQueue if available == 0 => Left(nonEmptyQueue)
-          }
-
-        val updated: State = old match {
-          case Left(waiting) =>
-            releaseRecursively(waiting, toRelease)
-          case Right(available) =>
-            Right(available + toRelease)
-        }
-
-        ((old, updated), updated)
-      }.flatMap {
-        case (previous, now) =>
-          previous match {
-            case Left(neq) =>
-              val newSize = now match {
-                case Left(newNeq) => newNeq.size
-                case Right(_)     => 0
-              }
-              val released = neq.size - newSize
-              neq.take(released).foldRight(IO.unit) { (entry, unit) =>
-                openGate(entry) *> unit
-              }
-            case Right(_) => IO.unit
-          }
-      }
-
-  private def mkGate: IO[Nothing, Promise[Nothing, Unit]] = Promise.make[Nothing, Unit]
-
-  private def awaitGate(entry: (Long, Promise[Nothing, Unit])): IO[Nothing, Unit] =
-    IO.unit.bracket0[Nothing, Unit] { (_, useOutcome) =>
-      useOutcome match {
-        case _: Terminated[Nothing, Unit] => // Terminated outcome means either interruption or uncaught exception
-          state.update {
-            case Left((current, rest)) =>
-              // if entry is NonEmpty's head and queue is empty, swap to Right, but without any permits
-              if (current == entry && rest.isEmpty) Right(0)
-              // if entry is NonEmpty's head and queue is not empty, just drop this entry from head position
-              else if (current == entry && rest.nonEmpty) Left(rest.dequeue)
-              // this entry is not current NonEmpty's head, just drop it from queue
-              else Left((current, rest.filter(_ != entry)))
-
-            case Right(m) => Right(m)
-          }.void
-        case _ =>
-          IO.unit
-      }
-    }(_ => entry._2.get)
-
-  private def openGate[E](entry: (Long, Promise[E, Unit])): IO[E, Unit] =
-    entry._2.complete(()).void
-
-  private def count_(state: State): Long = state match {
-    case Left((head, tail)) => -(head._1 + tail.map(_._1).sum)
-    case Right(available)   => available
+  final def acquireN(requested: Long): IO[Nothing, Unit] = {
+    val acquire: (Promise[Nothing, Unit], State) => (IO[Nothing, Boolean], State) = {
+      case (p, Right(n)) if n >= requested => p.complete(()) -> Right(n - requested)
+      case (p, Right(n))                   => IO.now(false)  -> Left(IQueue(p -> (requested - n)))
+      case (p, Left(q))                    => IO.now(false)  -> Left(q.enqueue(p -> requested))
+    }
+
+    val release: (Boolean, Promise[Nothing, Unit]) => IO[Nothing, Unit] = {
+      case (_, p) =>
+        p.poll.void <> state.update {
+          case Left(q) => Left(q.filterNot(_._1 == p))
```

Comment:
Yes, this will help ensure the promise is removed. I'd use `eq` instead of `==`.

Reply:
We'd have to make `Promise` a subtype of `AnyRef`, which sounds reasonable anyway?

Reply:
Nah, let's just leave it.